packages/relay/src/lib/services/metricService/metricService.ts

```diff
@@ -0,0 +1,366 @@
+   * @type {HbarLimit}
+   * @readonly
+   * @private
+   */
+  private readonly hbarLimiter: HbarLimit;
+
+  /**
+   * Histogram for capturing the cost of transactions and queries.
+   * @type {Histogram}
+   * @readonly
+   * @private
+   */
+  private readonly consensusNodeClientHistogramCost: Histogram;
+
+  /**
+   * Histogram for capturing the gas fee of transactions and queries.
+   * @type {Histogram}
+   * @readonly
+   * @private
+   */
+  private readonly consensusNodeClientHistogramGasFee: Histogram;
+
+  /**
+   * An instance of EventEmitter used for emitting and handling events within the class.
+   *
+   * @private
+   * @readonly
+   * @type {EventEmitter}
+   */
+  private readonly eventEmitter: EventEmitter;
+
+  /**
+   * Constructs an instance of the MetricService responsible for tracking and recording various metrics
+   * related to Hedera network interactions and resource usage.
+   *
+   * @param {Logger} logger - Logger instance for logging system messages.
+   * @param {SDKClient} sdkClient - Client for interacting with the Hedera SDK.
+   * @param {MirrorNodeClient} mirrorNodeClient - Client for querying the Hedera mirror node.
+   * @param {HbarLimit} hbarLimiter - Rate limiter for managing HBAR-related operations.
+   * @param {Registry} register - Registry instance for registering metrics.
+   * @param {EventEmitter} eventEmitter - The eventEmitter used for emitting and handling events within the class.
+   */
+  constructor(
+    logger: Logger,
+    sdkClient: SDKClient,
+    mirrorNodeClient: MirrorNodeClient,
+    hbarLimiter: HbarLimit,
+    register: Registry,
+    eventEmitter: EventEmitter,
+  ) {
+    this.logger = logger;
+    this.sdkClient = sdkClient;
+    this.hbarLimiter = hbarLimiter;
+    this.eventEmitter = eventEmitter;
+    this.mirrorNodeClient = mirrorNodeClient;
+    this.consensusNodeClientHistogramCost = this.initCostMetric(register);
+    this.consensusNodeClientHistogramGasFee = this.initGasMetric(register);
+
+    //listen to START_CAPTURING_TRANSACTION_METRICS event to kick off captureTransactionMetrics() process
+    this.eventEmitter.on(constants.EVENTS.START_CAPTURING_TRANSACTION_METRICS, (args) => {
+      this.captureTransactionMetrics(
+        args.transactionId,
+        args.callerName,
+        args.requestId,
+        args.txConstructorName,
+        args.operatorAccountId,
+        args.transactionType,
+        args.interactingEntity,
+      );
+    });
+
+    //listen to START_ADD_EXPENSE_AND_CAPTURE_METRICS event to kick off addExpenseAndCaptureMetrics() process
+    this.eventEmitter.on(constants.EVENTS.START_ADD_EXPENSE_AND_CAPTURE_METRICS, (args) => {
+      this.addExpenseAndCaptureMetrics(
+        args.executionType,
+        args.transactionId,
+        args.transactionType,
+        args.callerName,
+        args.cost,
+        args.gasUsed,
+        args.interactingEntity,
+        args.formattedRequestId,
+      );
+    });
+  }
```

Comment:
The `SdkClient` should emit the `TRANSACTION_EXECUTED` event since it isn’t aware of `MetricService`’s existence. `MetricService` can then subscribe to this event and handle it as needed. Having separate events for each method in `MetricService` and emitting them inside `SdkClient` is counterproductive - it’s similar to making `MetricService` a dependency, which defeats the purpose of using an `EventEmitter`:
```typescript
    this.eventEmitter.on(constants.TRANSACTION_EXECUTED, (args) => {
      // capture metrics for grafana dashboard
      this.captureTransactionMetrics(
        args.transactionId,
        args.callerName,
        args.requestId,
        args.txConstructorName,
        args.operatorAccountId,
        args.transactionType,
        args.interactingEntity,
      );
      // add expense to hbar limiter spending
      this.addExpenseAndCaptureMetrics(
        args.executionType,
        args.transactionId,
        args.transactionType,
        args.callerName,
        args.cost,
        args.gasUsed,
        args.interactingEntity,
        args.formattedRequestId,
      );
    });
  }
```

Reply:
Also, you can change `captureTransactionMetrics` and `addExpenseAndCaptureMetrics` to accept an object argument (you can use object de-structuring in the method's arguments) and change this to:
```typescript
    this.eventEmitter.on(constants.TRANSACTION_EXECUTED, (args) => {
      // capture metrics for grafana dashboard
      this.captureTransactionMetrics(args);
      // add expense to hbar limiter spending
      this.addExpenseAndCaptureMetrics(args);
    });
  }
```

Reply:
> The `SdkClient` should emit the `TRANSACTION_EXECUTED` event since it isn’t aware of `MetricService`’s existence. `MetricService` can then subscribe to this event and handle it as needed. Having separate events for each method in `MetricService` and emitting them inside `SdkClient` is counterproductive - it’s similar to making `MetricService` a dependency, which defeats the purpose of using an `EventEmitter`:
> 
> ```ts
>     this.eventEmitter.on(constants.TRANSACTION_EXECUTED, (args) => {
>       // capture metrics for grafana dashboard
>       this.captureTransactionMetrics(
>         args.transactionId,
>         args.callerName,
>         args.requestId,
>         args.txConstructorName,
>         args.operatorAccountId,
>         args.transactionType,
>         args.interactingEntity,
>       );
>       // add expense to hbar limiter spending
>       this.addExpenseAndCaptureMetrics(
>         args.executionType,
>         args.transactionId,
>         args.transactionType,
>         args.callerName,
>         args.cost,
>         args.gasUsed,
>         args.interactingEntity,
>         args.formattedRequestId,
>       );
>     });
>   }
> ```

@victor-yanev Great idea! However, `START_CAPTURING_TRANSACTION_METRICS` and `START_ADD_EXPENSE_AND_CAPTURE_METRICS` are two distinct events that require separate flows. Combining them into a single flow wouldn’t be accurate, as some functions only require one event or the other, but not both. If anything, should we consider adding an `args.type` and using switch statements to distinguish between the flows?


for example 
```typescript
    this.eventEmitter.on(constants.TRANSACTION_EXECUTED, (args) => {
     switch (args.type) {
          case "abc":
                // capture metrics for grafana dashboard
               this.captureTransactionMetrics(
                   args.transactionId,
                   args.callerName,
                   args.requestId,
                   args.txConstructorName,
                   args.operatorAccountId,
                   args.transactionType,
                   args.interactingEntity,
              );
              break;
              
          case "xyz":
               // add expense to hbar limiter spending
               this.addExpenseAndCaptureMetrics(
                   args.executionType,
                   args.transactionId,
                   args.transactionType,
                   args.callerName,
                   args.cost,
                   args.gasUsed,
                   args.interactingEntity,
                   args.formattedRequestId,
      );
    });        
     }
  }
```

Reply:
In which functions do we require only one? Could you clarify?

I think we should always capture any hbar spending using the `captureTransactionMetrics`, as well as add it to the hbar limiter's expenses using `addExpenseAndCaptureMetrics`